exercises/circular-buffer/tests/circular-buffer.rs

```diff
@@ -1,113 +1,121 @@
 extern crate circular_buffer;
+use circular_buffer::{CircularBuffer, Error};
 
-#[allow(unused_must_use)]
-mod tests {
-
-    use circular_buffer::{CircularBuffer, Error};
-
-    #[test]
-    fn error_on_read_empty_buffer() {
-        let mut buffer = CircularBuffer::<char>::new(1);
-        assert_eq!(Err(Error::EmptyBuffer), buffer.read());
-    }
+#[test]
+fn error_on_read_empty_buffer() {
+    let mut buffer = CircularBuffer::<char>::new(1);
+    assert_eq!(Err(Error::EmptyBuffer), buffer.read());
+}
 
-    #[test]
-    #[ignore]
-    fn write_and_read_back_item() {
-        let mut buffer = CircularBuffer::new(1);
-        buffer.write('1');
-        assert_eq!(Ok('1'), buffer.read());
-        assert_eq!(Err(Error::EmptyBuffer), buffer.read());
-    }
+#[test]
+#[ignore]
+fn write_and_read_back_item() {
+    let mut buffer = CircularBuffer::new(1);
+    // Normally, using `unwrap` like this would be pretty bad practice:
+    // the call panics on error, which is bad behavior on the part of an
+    // executable or library.
+    //
+    // It's less of a problem within a test, for two reasons:
+    //  1. We control the entire test environment, and so should be able to
+    //     predict exactly when errors will and will not occur.
+    //  2. A panic here simply fails the test instead of aborting in the middle
+    //     of unrelated code.
+    //
+    // Finally, it underscores the fact that this isn't actually the behavior
+    // under test at the moment: when we want to test some behavior, we use assert.
+    // `unwrap()`, in this context, means that we're just trying to get to the point.
+    buffer.write('1').unwrap();
```

Comment:
This is the very first time `write` is encountered in this test file, and the function even has `write` in its name. That makes me wonder about the "this isn't actually the behavior under test" part of the comment.

The previous situation was that the `write` was being tested, but only indirectly: If the `write` had failed unexpectedly, we would see the subsequent `read` return an unexpected value.

Now, we are testing that the `write` succeeds, as we expect it to. I think this is better: It is good to notify as early as possible when we know there are failures.

If this is not a situation where we consider `write` part of the behaviour under test... under what kind of situation would we actually consider it to be?

I suppose my own answer to this question would be: a test entitled "full buffer can't be written to" as in https://github.com/exercism/problem-specifications/blob/master/exercises/circular-buffer/canonical-data.json#L116-L133, or something similar.

Reply:
I see. I think I'm discovering a preference for `assert!(buffer.write(...).is_ok())`, because it's very explicit about what we expect, and it means that we can eliminate the entire section of comments justifying the use of unwrap.

Reply:
I'm good with that as well.